File: src/lexer/lexer.cpp
#include <iostream>
#include <regex>

enum TokenType
{
    PLUS,
    MINUS,
    MULTIPLY,
    INTEGER,
    FLOAT,
    STR,
    BOOL,
    EQUALITY,
    INEQUALITY,
    GREATER,
    LESS,
    AND,
    OR,
    NOT,
    IF,
    ELSEIF,
    ELSE,
    FOR,
    WHILE,
    LOOP,
    ENDLOOP,
    CONTINUE,
    BREAK,
    STRUCT,
    ENUM,
    FUNC,
    RETURNTYPE,
    RETURN,
    NONE,
    NEWLINE,
    DIVIDE,
    MODULO,
    LPAREN,
    SPACE,
    NUMBER,
    IDENTIFIER,
    RPAREN,
    HASHTAG,
    SEMICOLON,
    LESSEQUAL,
    GREATEREQUAL,
    EQUAL,
    COMMENT,
    SLASH,
    SINGLEQUOTE,
    DOUBLEQUOTE,
    COMMA,
    DOT,
    COLON,
    LCURLY,
    RCURLY,
    LSQUARE,
    RSQUARE,
    STRING,
    INCREMENT,
    DECREMENT,
    ADDITIONASSIGNMENT,
    SUBTRACTIONASSIGNMENT,
    MULTIPLICATIONASSIGNMENT,
    DIVISIONASSIGNMENT,
    MODULUSASSIGNMENT,
    IMPORT,
    END_OF_FILE
};
struct Token
{
    TokenType type;
    std::string value;
};

const char* tokens[]{ "PLUS" , "MINUS", "MULTIPLY","INTEGER","FLOAT","STRING","BOOL","EQUALITY",
                      "INEQUALITY","GREATER","LESS","AND","OR","NOT","IF","ELSEIF","ELSE","FOR",
                      "WHILE","LOOP","ENDLOOP","CONTINUE","BREAK","STRUCT","ENUM","FUNCTION",
                      "RETURNTYPE","RETURN","NONE","NEWLINE", "DIVIDE","MODULO", "LPAREN", "SPACE", "NUMBER",
                      "IDENTIFIER", "RPAREN","HASHTAG","SEMICOLON","LESSEQUAL","GREATEREQUAL",
                      "EQUAL","COMMENT","SLASH","SINGLEQUOTE","DOUBLEQUOTE","COMMA","DOT",
                      "COLON","LCURLY","RCURLY","LSQUARE","RSQUARE","STRING","INCREMENT",
                      "DECREMENT","ADDITIONASSIGNMENT","SUBTRACTIONASSIGNMENT","MULTIPLICATIONASSIGNMENT",
                      "DIVISIONASSIGMENT","MODULUSASSIGNMENT","IMPORT", "END_OF_FILE" };
class Lexer
{
public:
    Lexer(const std::string& text) : text(text), pos(0) {}
    Token getNextToken()
    {
        if (pos >= text.length())
        {
            return { END_OF_FILE, "" };
        }
        char currentChar = text[pos];
        if (isdigit(currentChar))
        {
            return isDigit();
        }


        if (isalpha(currentChar))
        {
            return isLetter();
        }
        switch (currentChar)
        {
        case '+':
            pos++;
            if (getNextChar() == '+') { pos++; return { INCREMENT, "++" }; break; }
            else if (getNextChar() == '=') { pos++; return { ADDITIONASSIGNMENT, "+=" }; break; }
            return { PLUS, "+" };
            break;

        case '-':
            pos++;
            if (getNextChar() == '-') { pos++; return { DECREMENT, "--" }; break; }
            else if (getNextChar() == '=') { pos++; return { SUBTRACTIONASSIGNMENT, "-=" }; break; }
            return { MINUS, "-" };
            break;

        case '/':
            pos++;
            if (getNextChar() == '=') { pos++; return { DIVISIONASSIGNMENT, "/=" }; break; }
            return { DIVIDE, "/" };
            break;
        case '%':
            pos++;
            if (getNextChar() == '=') { pos++; return { MODULUSASSIGNMENT, "%=" }; break; }
            return { MODULO, "%" };
            break;

        case '*':
            pos++;
            if (getNextChar() == '=') { pos++; return { MULTIPLICATIONASSIGNMENT, "%=" }; break; }
            return { MULTIPLY, "*" };
            break;

        case '(':
            pos++;
            return { LPAREN, "(" };
            break;

        case ')':
            pos++;
            return { RPAREN, ")" };
            break;

        case ' ':
            pos++;
            return { SPACE, " " };
            break;
        case '\n':
            pos++;
            return { NEWLINE, "\\n" };
            break;
        case ';':
            pos++;
            return { SEMICOLON, ";" };
            break;
        case '<':
            pos++;
            if (getNextChar() == '=')
            {
                pos++;
                return { LESSEQUAL, "<=" };
                break;
            }

            return { LESS, "<" };
            break;
        case '>':
            pos++;
            if (getNextChar() == '=')
            {
                pos++;
                return { GREATEREQUAL, "<=" };
                break;
            }
            return { GREATER, ">" };
            break;
        case '=':
            pos++;
            if (getNextChar() == '=')
            {
                pos++;
                return { EQUALITY, "==" };
                break;
            }
            else if (getNextChar() == '!')
            {
                pos++;
                return { INEQUALITY, "=!" };
                break;
            }
            return { EQUAL, "=" };
            break;
        case '\'':
            pos++;
            return { SINGLEQUOTE, "'" };
            break;
        case '\"':
            pos++;
            return LiteralString();
            break;
        case '.':
            pos++;
            return { DOT, "." };
            break;
        case ',':
            pos++;
            return { COMMA, "," };
            break;
        case '{':
            pos++;
            return { LCURLY, "{" };
            break;
        case '}':
            pos++;
            return { RCURLY, "}" };
            break;
        case ':':
            pos++;
            return { COLON, ":" };
            break;
        case '[':
            pos++;
            return { LSQUARE, "[" };
            break;
        case ']':
            pos++;
            return { RSQUARE, "]" };
        case '#':
            return isComment();
            break;
        }

    }
private:
    char getNextChar()
    {
        char nextChar = text[pos];
        return nextChar;
    }
    Token Identifier(const std::string& text)
    {
        if (text == "int") return { INTEGER, "INTEGER" };
        else if (text == "float") return { FLOAT, "FLOAT" };
        else if (text == "string") return { STR, "STR" };
        else if (text == "bool") return { BOOL, "BOOLEAN" };
        else if (text == "and") return { AND, "AND" };
        else if (text == "or") return { OR, "OR" };
        else if (text == "not") return { NOT, "NOT" };
        else if (text == "if") return { IF, "IF" };
        else if (text == "elseif") return { ELSEIF, "ELSEIF" };
        else if (text == "else") return { ELSE, "ELSE" };
        else if (text == "for") return { FOR, "FOR" };
        else if (text == "while") return { WHILE, "WHILE" };
        else if (text == "continue") return { CONTINUE, "CONTINUE" };
        else if (text == "break") return { BREAK, "BREAK" };
        else if (text == "struct") return { STRUCT, "STRUCT" };
        else if (text == "enum") return { ENUM, "ENUM" };
        else if (text == "fn") return { FUNC, "FUNCTION" };
        else if (text == "return") return { RETURN, "RETURN" };
        else if (text == "none") return { NONE, "NULL" };
        else if (text == "true") return { BOOL, "TRUE" };
        else if (text == "false") return { BOOL, "FALSE" };
        else if (text == "include") return { IMPORT, "INCLUDE" };
        else return { IDENTIFIER, text };
    }
    Token isDigit()
    {
        std::string result = "";
        while (pos <= text.length() && isdigit(text[pos]))
        {
            result += text[pos];
            pos++;
        }
        return { NUMBER, result };
    }
    Token isLetter()
    {
        std::string result = "";
        while (pos <= text.length() && isalpha(text[pos]))
        {
            result += text[pos];
            pos++;
        }
        return Identifier(result);
    }
    Token LiteralString()
    {
        std::string result = "";
        while (pos <= text.length() && text[pos] != '\"')
        {
            result += text[pos];
            pos++;
        }
        pos++;
        if (result == "\"") return { DOUBLEQUOTE, "\"" };
        return { STRING, result };
    }
    Token isComment()
    {
        std::string result = "";
        while (pos <= text.length() && text[pos] == '#')
        {
            result += text[pos];
            pos++;
        }
        if (result == "##")
        {
            result = "";
            while (pos <= text.length() && text[pos] != '\n')
            {
                result += text[pos];
                pos++;
            }
            return { COMMENT, result };
        };
        return { HASHTAG, result };
    }
    std::string text = "";
    size_t pos;

};

int main()
{
    auto input = "fn merhaba(x,y):\n"
        "int test = 30\n"
        "return test+30\n"
        "string merhaba = \"test selam not or false merhaba and and and =>\"\n"
        "bool merhaba = true;\n"
        "for(int x = 0; x <= 30; x++)\n"
        "if(x == test): return true\n"
        "print(\"Sifirdan programlama dili gelistiriyoruz.\")\n"
        "x++;"
        "y+=30;"
        "c %= 2"
        "if(c % 2 == 0) return true;"
        "while (true): break;"
        "include discord";

    /*std::regex pattern("\\s+");
    std::string result = std::regex_replace(input, pattern, "");*/
    Lexer lexer(input);
    Token token = lexer.getNextToken();
    while (token.type != END_OF_FILE)
    {
        if (token.type != SPACE) std::cout << tokens[token.type] << " |" << token.value << "| " << std::endl;
        token = lexer.getNextToken();
    }
    std::cout << tokens[TokenType::END_OF_FILE];
    return 0;
}

File: src/lexer/lexer.cs
using System;

namespace HelloWorld
{
    public enum TokenType
    {
        Plus,
        Minus,
        Multiply,
        Integer,
        Float,
        String,
        Bool,
        Equality,
        Inequality,
        Greater,
        Less,
        And,
        Or,
        Not,
        If,
        Elif,
        Else,
        For,
        While,
        Loop,
        EndLoop,
        Continue,
        Break,
        Enum,
        Function,
        Return,
        ReturnType,
        None,
        NewLine,
        Divide,
        Modulo,
        LParen,
        RParen,
        Space,
        Number,
        Text,
        Identifier,
        Semicolon,
        LessEqual,
        GreaterEqual,
        Equal,
        Comment,
        Slash,
        SingleQuote,
        DoubleQuote,
        Comma,
        Dot,
        Colon,
        LCurly,
        RCurly,
        LSquare,
        RSquare,
        Increment,
        Decrement,
        AdditionAssignment,
        SubtractionAssignment,
        MultiplicationAssignment,
        DivisionAssignment,
        ModuloAssignment,
        Hashtag,
        Other,
        End_Of_File
    }

    public struct Token
    {
        public TokenType type;
        public string value;

        public Token(TokenType tokenType, string tokenValue)
        {
            type = tokenType;
            value = tokenValue;
        }
    }

    public class Lexer
    {
        private string text;
        private int pos;

        public Lexer(string lexText)
        {
            text = lexText;
            pos = 0;
        }

        public Token getNextToken()
        {
            if (pos >= text.Length)
            {
                return new Token(TokenType.End_Of_File, "END");
            }

            char currentChar = text[pos];
            if (Char.IsDigit(currentChar))
            {
                return isDigit();
            }


            if (Char.IsLetter(currentChar))
            {
                return isLetter();
            }

            switch (currentChar)
            {
                case '+':
                    pos++;
                    if (getNextChar() == '+')
                    {
                        pos++;
                        return new Token(TokenType.Increment, "++");
                        break;
                    }
                    else if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.AdditionAssignment, "+=");
                        break;
                    }

                    return new Token(TokenType.Plus, "+");
                    break;

                case '-':
                    pos++;
                    if (getNextChar() == '-')
                    {
                        pos++;
                        return new Token(TokenType.Decrement, "--");
                        break;
                    }
                    else if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.SubtractionAssignment, "-=");
                        break;
                    }

                    return new Token(TokenType.Minus, "-");
                    break;

                case '/':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.DivisionAssignment, "/=");
                        break;
                    }

                    return new Token(TokenType.Divide, "/");
                    break;
                case '%':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.ModuloAssignment, "%=");
                        break;
                    }

                    return new Token(TokenType.Modulo, "%");
                    break;
                    break;

                case '*':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.MultiplicationAssignment, "%=");
                        break;
                    }

                    return new Token(TokenType.Multiply, "*");
                    break;

                case '(':
                    pos++;
                    return new Token(TokenType.LParen, "(");
                    break;

                case ')':
                    pos++;
                    return new Token(TokenType.RParen, ")");
                    break;

                case ' ':
                    pos++;
                    return new Token(TokenType.Space, " ");
                    break;
                case '\n':
                    pos++;
                    return new Token(TokenType.NewLine, "\\n");
                    break;
                case ';':
                    pos++;
                    return new Token(TokenType.Semicolon, ";");
                    break;
                case '<':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.LessEqual, "<=");
                        break;
                    }

                    return new Token(TokenType.Less, "<");
                    break;
                case '>':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.GreaterEqual, "<=");
                        break;
                    }

                    return new Token(TokenType.Greater, ">");
                    break;
                case '=':
                    pos++;
                    if (getNextChar() == '=')
                    {
                        pos++;
                        return new Token(TokenType.Equality, "==");
                        break;
                    }
                    else if (getNextChar() == '!')
                    {
                        pos++;
                        return new Token(TokenType.Inequality, "=!");
                        break;
                    }

                    return new Token(TokenType.Equality, "=");
                    break;
                case '\'':
                    pos++;
                    return new Token(TokenType.SingleQuote, "'");
                    break;
                case '\"':
                    pos++;
                    return LiteralString();
                    break;
                case '.':
                    pos++;
                    return new Token(TokenType.Dot, ".");
                    break;
                case ',':
                    pos++;
                    return new Token(TokenType.Comma, ",");
                    break;
                case '{':
                    pos++;
                    return new Token(TokenType.LCurly, "{");
                    break;
                case '}':
                    pos++;
                    return new Token(TokenType.RCurly, "}");
                    break;
                case '[':
                    pos++;
                    return new Token(TokenType.LSquare, "[");
                    break;
                case ']':
                    pos++;
                    return new Token(TokenType.RSquare, "]");
                    break;
                case '#':
                    return isComment();
                    break;
                case ':':
                    pos++; 
                    return new Token(TokenType.Colon, ":");
                    break;
                default:
                    pos++; 
                    return new Token(TokenType.Other, currentChar.ToString());
                    break;
            }
        }

        public char getNextChar()
        {
            char nextChar = text[pos];
            return nextChar;
        }

        public Token Identifier(string text)
        {
            if (text == "int") return new Token(TokenType.Integer, "INTEGER");
            else if (text == "float") return new Token(TokenType.Float, "FLOAT");
            else if (text == "str") return new Token(TokenType.String, "STR");
            else if (text == "bool") return new Token(TokenType.Bool, "BOOLEAN");
            else if (text == "and") return new Token(TokenType.And, "AND");
            else if (text == "or") return new Token(TokenType.Or, "OR");
            else if (text == "not") return new Token(TokenType.Not, "NOT");
            else if (text == "if") return new Token(TokenType.If, "IF");
            else if (text == "elif") return new Token(TokenType.Elif, "ELSEIF");
            else if (text == "else") return new Token(TokenType.Else, "ELSE");
            else if (text == "for") return new Token(TokenType.For, "FOR");
            else if (text == "while") return new Token(TokenType.While, "WHILE");
            else if (text == "continue") return new Token(TokenType.Continue, "CONTINUE");
            else if (text == "break") return new Token(TokenType.Break, "BREAK");
            else if (text == "fn") return new Token(TokenType.Function, "FUNCTION");
            else if (text == "ret") return new Token(TokenType.Return, "RETURN");
            else if (text == "none") return new Token(TokenType.None, "NULL");
            else if (text == "true") return new Token(TokenType.Bool, "TRUE");
            else if (text == "false") return new Token(TokenType.Bool, "FALSE");
            else return new Token(TokenType.Identifier, text);
        }

        public Token isDigit()
        {
            string result = "";
            while (pos <= text.Length && Char.IsDigit(text[pos]))
            {
                result += text[pos];
                pos++;
            }

            return new Token(TokenType.Number, result);
        }

        public Token isLetter()
        {
            string result = "";
            while (pos <= text.Length && Char.IsLetter(text[pos]))
            {
                result += text[pos];
                pos++;
            }

            return Identifier(result);
        }

        public Token LiteralString()
        {
            string result = "";
            while (pos < text.Length && text[pos] != '\"')
            {
                result += text[pos];
                pos++;
            }

            pos++;
            if (result == "\"") return new Token(TokenType.DoubleQuote, "\"");
            return new Token(TokenType.String, result);
        }

        public Token isComment()
        {
            string result = "";
            while (pos <= text.Length && text[pos] == '#')
            {
                result += text[pos];
                pos++;
            }

            if (result == "##")
            {
                result = "";
                while (pos <= text.Length && text[pos] != '\n')
                {
                    result += text[pos];
                    pos++;
                }

                return new Token(TokenType.Comment, result);
            }

            ;
            return new Token(TokenType.Hashtag, result);
        }
    }

    class Program
    {
        public static string[] tokens = {
            "PLUS", "MINUS", "MULTIPLY", "INTEGER", "FLOAT", "STRING", "BOOL", "EQUALITY",
            "INEQUALITY", "GREATER", "LESS", "AND", "OR", "NOT", "IF", "ELIF", "ELSE", "FOR",
            "WHILE", "LOOP", "ENDLOOP", "CONTINUE", "BREAK", "ENUM",
            "FUNCTION", "RETURN", "RETURNTYPE", "NONE", "NEWLINE", "DIVIDE", "MODULO", "LPAREN", "RPAREN", "SPACE", "NUMBER",
            "TEXT", "IDENTIFIER", "SEMICOLON", "LESSEQUAL", "GREATEREQUAL",
            "EQUAL", "COMMENT", "SLASH", "SINGLEQUOTE", "DOUBLEQUOTE", "COMMA", "DOT","COLON",
            "LCURLY", "RCURLY", "LSQUARE", "RSQUARE", "INCREMENT",
            "DECREMENT", "ADDITIONASSIGNMENT", "SUBTRACTIONASSIGNMENT", "MULTIPLICATIONASSIGNMENT",
            "DIVISIONASSIGNMENT", "MODULOASSIGNMENT", "HASHTAG", "OTHER", "END_OF_FILE"
        };

        static void Main(string[] args)
        {
            string input =
                "fn merhaba(x,y){\nint test = 30;\nret test+30}\nstr merhaba = \"test selam not or false merhaba and and and =>\"\nbool merhaba = true;\nfor(int x = 0; x <= 30; x++){break;}\nif(x == test){ return true;}\nwrite(\"Sifirdan programlama dili gelistiriyoruz.\");\nx++;\ny+=30;\nc %= 2;\nif(c % 2 == 0){ return true;}\nwhile (true){ break;}";
            Lexer lexer = new Lexer(input);
            Token token;
            do
            {
                token = lexer.getNextToken();
                if (token.type != TokenType.Space)
                    Console.WriteLine($"{tokens[(int)token.type]} | {token.value} |");
            } while (token.type != TokenType.End_Of_File);
        }
    }
}